File: WorldFeed/Server/src/WorldFeed.Common/DTO/Cursor/MessageCursorQueryResultDTO.cs
namespace WorldFeed.Common.DTO.Cursor
{
    using System.Collections.Generic;
    using Newtonsoft.Json;

    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Models.Interfaces.DTO.Events;
    using WorldFeed.Common.Public.Models.Interfaces.DTO.QueryDTO;

    public class MessageCursorQueryResultDTO : BaseCursorQueryDTO<IMessageEventDTO>, IMessageCursorQueryResultDTO
    {
        private IMessageEventDTO[] messageEvents;

        // This property does not exists
        [JsonIgnore]
        public override long NextCursor { get; set; }

        [JsonProperty("next_cursor")]
        public override string NextCursorStr { get; set; }

        public override int GetNumberOfObjectRetrieved()
        {
            return MessageEvents.Length;
        }

        [JsonProperty("events")]
        public IMessageEventDTO[] MessageEvents
        {
            get => this.messageEvents ?? new IMessageEventDTO[0];
            set
            {
                this.messageEvents = value;
                Results = value;
            }
        }

        [JsonProperty("apps")]
        public Dictionary<long, IApp> Apps { get; set; }
    }
}

File: Sharebook/Client/src/app/core/Core/DTO/Cursor/MessageCursorQueryResultDTO.ts
import {BaseCursorQueryDTO} from "./BaseCursorQueryDTO";
import {IApp} from "../../../Public/Models/Interfaces/IApp";
import {IMessageEventDTO} from "../../../Public/Models/Interfaces/DTO/Events/IMessageEventDTO";
import {IMessageCursorQueryResultDTO} from "../../../Public/Models/Interfaces/DTO/QueryDTO/IMessageCursorQueryResultDTO";
import {MessageEventDTO} from "../Events/MessageEventDTO";
import Dictionary from "typescript-dotnet-commonjs/System/Collections/Dictionaries/Dictionary";

export class MessageCursorQueryResultDTO extends BaseCursorQueryDTO<IMessageEventDTO> implements IMessageCursorQueryResultDTO {
  private _messageEvents: IMessageEventDTO[];

  // This property does not exists
  // [JsonIgnore]
  public nextCursor: number;   // override

  // [JsonProperty("next_cursor")]
  public nextCursorStr: string;

  // [JsonProperty("apps")]
  public apps: Dictionary<number, IApp>;

  public getNumberOfObjectRetrieved(): number {
    return this.messageEvents.length;
  }

  // [JsonProperty("events")]
  get messageEvents(): IMessageEventDTO[] {
    return this._messageEvents ?? new MessageEventDTO()[0];  // new IMessageEventDTO[0];
  }

  set messageEvents(value: IMessageEventDTO[]) {
    this._messageEvents = value;
    super.results = value;
  }
}
